python/runtime/cudaq/algorithms/py_state.cpp

```diff
@@ -45,64 +36,113 @@ state pyGetState(kernel_builder<> &kernel, py::args args) {
 /// @brief Bind the get_state cudaq function
 void bindPyState(py::module &mod) {
 
+  // A cudaq.State instance can not be constructed from Python code.
+  // It should only be constructed from an internal simulator, and that
+  // handle to the data (the cudaq.State instance) is returned to the user.
+  // State-relevant data can be generated from a cudaq.State with
+  // pertinent cudaq.State methods. (e.g. the overlap with an existing state
+  // reference). These methods should first and foremost operate on
+  // cudaq.State references. For ease of use, we also allow user-specified
+  // py::buffer_info adherent types (e.g. Numpy arrays). In the case where
+  // a py::buffer_info type is provided, care must be taken to ensure the data
+  // is made available in the right memory space (e.g. transferred from host
+  // to GPU memory). We should also allow a user to leverage GPU device memory
+  // for these types of computations, with something like CuPy.
+
   py::class_<state>(
       mod, "State", py::buffer_protocol(),
       "A data-type representing the quantum state of the internal simulator. "
       "Returns a state vector by default. If the target is set to "
-      "`density-matrix-cpu`, "
-      "a density matrix will be returned.\n")
-      .def_buffer([](state &self) -> py::buffer_info {
+      "`density-matrix-cpu`, a density matrix will be returned. This type is "
+      "not user-constructible and instances can only be retrieved via the "
+      "`cudaq.get_state(...)` function. \n")
+      .def(
+          py::init([](const py::buffer &b) -> cudaq::state {
+            throw std::runtime_error(
+                "cudaq.State can only be constructed from an internal "
+                "simulator. "
+                "State comparison operations (e.g. overlap) take Python buffer "
+                "data (e.g. NumPy array) as input by default, therefore there "
+                "is "
+                "no need to construct a cudaq.State from data buffers "
+                "directly.");
+          }),
+          R"#(`State` is not constructible from user code. Throw an error.)#")
+      .def_buffer([](const state &self) -> py::buffer_info {
+        // This method is used by Pybind to enable interoperability
+        // with NumPy array data. We therefore must be careful since the
+        // state data may actually be on GPU device.
+
+        // Get the data pointer.
+        // Data may be on GPU device, if so we must make a copy to host.
+        // If users do not want this copy, the will have to operate apart from
+        // Numpy
```

Comment:
@bmhowe23 @bettinaheim Should I throw a warning here in the case that a user tries to use a GPU-device state with NumPy, indicating there is an explicit copy of the data?

Reply:
Will the `cudaq::info` message in lines 96-97 eventually be triggered? If so, I think that is sufficient.